### ISSUE TITLE
Command group (application lifecycle)

### ISSUE THREAD

**Issue by kadel** [devfile/kubernetes-api#27]
Commands should have a `group` field that will specify in what point of application lifecycle the command should be executed (or type of the command like, build, incremental-build, run etc..). 

- `group` field is optional
- Many commands of the same devfile can belong to the same `group`
- Only one command belonging to a group can be the default 
- The list of groups should be: `build`, `run`, `test`, `debug`

```yaml
commands:
 - exec:
     name: "package"
     commandLine: "mvn package"
     component: build-tools
     group: 
        kind: build
 - exec:
     name: "install"
     commandLine: "mvn install"
     component: build-tools
     group: 
         kind: build
         isDefault: true 
```

**Comment by gorkem** [comment]
For reference vscode uses `group` instead of `type`.

**Comment by davidfestal** [comment]
I also prefer `group` instead of `type`, because it seems that `type` name is a bit confusing due to its frequent use in the context of *sub-typing*, or polymorphism.

Here we're more thinking of partitioning / grouping Exec commands into several categories, which is why I like the `group` (or `category`) field name better.

**Comment by kadel** [comment]
> Here we're more thinking of partitioning / grouping Exec commands into several categories, which is why I like the `group` (or `category`) field name better.

Completely agree. `group` or `category` is far better than `type` which is already overloaded term.

**Comment by l0rd** [comment]
We reviewed this proposal and agreed on the following specs:
-  We should use `group` instead of `type` for this proposal
- `group` field is optional
- Many commands of the same devfile can belong to the same `group`

We started discussing whether we should have or not a default command for a given `group` like

```yaml
commands:
 - exec:
     name: "package"
     commandLine: "mvn package"
     component: build-tools
     type: build
 - exec:
     name: "install"
     commandLine: "mvn install"
     component: build-tools
     type: build
     isDefault: true 
```

or 

```yaml
commands:
 - exec:
     name: "package"
     commandLine: "mvn package"
     component: build-tools
     type: build
 - exec:
     name: "install"
     commandLine: "mvn install"
     component: build-tools
     type: build
  - groupDefaults:
       - group: "build"
          command: "install" 
```

**Comment by gorkem** [comment]
Is there a list of recognized 'group's?

**Comment by l0rd** [comment]
@gorkem we haven't discussed that but haven't finished to discuss this issue. We will continue next week.

**Comment by davidfestal** [comment]
I assume that, in the second example proposed in comment https://github.com/che-incubator/devworkspace-api/issues/27#issuecomment-590984574, adding a single `groupDefaults` element at the end of the list of `commands` would not be very easy to define, both in Go and in the derived JSON schema, since it's a completely different schema, and not really a `Command`.
Additionally afaik enforcing it to appear only once through the schema validation might be impossible at this location.

```yaml
commands:
 - exec:
     name: "package"
     commandLine: "mvn package"
     component: build-tools
     type: build
 - exec:
     name: "install"
     commandLine: "mvn install"
     component: build-tools
     type: build
 - groupsDefaults:
     - group: "build"
       command: "install" 
```

**Comment by l0rd** [comment]
Discussed and agreed

**Comment by groeges** [comment]
Is there any way to define different commandLines dependant on which Operating System is being used, specifically for `type:  init` which will be run on the users local system?

**Comment by l0rd** [comment]
@groeges given that the commands are supposed to be executed in a container (components are k8s/openshift/container objects) and not as a local processes, supporting multiple OSes doesn't look like a valid use case.

**Comment by groeges** [comment]
@l0rd I agree that most of the commands will be running within the container so a single script is OK, however I thought there was a `devInitLocal` being added (or something like that) which will run a command to initialise the project on the users local workspace and this is the case that may need to have multiple commandLines based on OS. If this is not being added then this is not an issue, but otherwise it is.

**Comment by l0rd** [comment]
@groeges we are discussing that in #30 but I consider that devfiles describe container-based development workspaces and making it suitable for local dev (i.e. running local processes, accessing local files) is out of scope. But we can discuss that.

**Comment by davidfestal** [comment]
Implemented
